lib/verdict/storage/cookie_storage.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+module Verdict
+  module Storage
+    # CookieStorage, unlike other Verdict storage classes, is distributed and stored on the client.
+    # Because cookies are opaque to users, we obsfucate information about the test such as the
+    # human readable names for the experiment or assignment group. This means this class assumes
+    # that `name` will be an obsfucated value, or one that we comfortable being "public".
+    class CookieStorage < BaseStorage
+      DEFAULT_COOKIE_LIFESPAN_SECONDS = 15778476 # 6 months
+
+      attr_accessor :cookies
+      attr_reader :cookie_lifespan
+
+      def initialize(options = {})
+        @cookies = options[:cookies] || {}
+        @cookie_lifespan = options[:cookie_lifespan] || DEFAULT_COOKIE_LIFESPAN_SECONDS
+      end
+
+      def store_assignment(assignment)
+        hash = { group: assignment.group.name, created_at: assignment.created_at.strftime('%FT%TZ') }
+        set("verdict_#{assignment.experiment.name}", nil, JSON.dump(hash))
+      end
+
+      def retrieve_assignment(experiment, subject)
+        if (value = get("verdict_#{experiment.name}", nil))
+          hash = parse_cookie_value(value)
+          group = experiment.groups.values.find { |g| g.name == hash['group'] }
+
+          if group.nil?
+            experiment.remove_subject_assignment(subject)
+            return nil
+          end
+
+          experiment.subject_assignment(subject, group, Time.xmlschema(hash['created_at']))
+        end
+      end
+
+      def remove_assignment(experiment, _subject)
+        remove("verdict_#{experiment.name}", nil)
+      end
+
+      def retrieve_start_timestamp(_experiment)
+        nil
+      end
+
+      def store_start_timestamp(_experiment, _timestamp)
+        true
```

Comment:
I assume this is still possible to support?

Reply:
Technically yes, but I didn't see the value. This is called via `ensure_experiment_has_started` which is called during assignment (`subject_qualifies?`). 

In practice with the storage being distributed, it'd be mean a fresh cookie is stored on every user's browser with the experiment name and started date, where the started date would be different for every user (it's set to `Time.now.utc`).

Reply:
Oh, I see what you mean. Yeah, I guess the fact that we are using a distributed datastore we can't have a centralized start time (unless we use a different datastore to store that, which might be overkill). We also can't raise an error to explain the distribution problem, but if we introduce a new error type and rescue it in `ensure_experiment_has_started`, that would work. That would also make our intent more clear, but I'll leave it up to you.

Reply:
This now raises a `NotImplementedError`, which is now handled in `Experiment.set_start_timestamp`.